Chrome tab style

DivineDominion opened the pull request:
I added a `ChromeStyle` to draw tabs with an angled left/right border and a visual top padding.

ℹ️ Increasing the bottom TabsControl height to about 34 looks way better. 

I also added `Style.tabButtonOffset` to let tab buttons overlap if the style wants to. (`ChromeStyle` does :))

The drawing code is kind of ugly. I started with a `PaddedHeight` enum but found it grew rather clunky in the process.

@onekiloparsec What do you think so far? 

---

(This is work in progress and not ready to be merged.)

DivineDominion committed:
prepare ChromeStyle

DivineDominion committed:
draw basic tab bezel

DivineDominion committed:
improve title positioning

DivineDominion committed:
draw tabs control bottom border

DivineDominion committed:
draw selected tab with brighter background

DivineDominion committed:
make bottom pane adopt Chrome BG color

- add ColoredView NSView subclass to set background color easily

DivineDominion committed:
draw bottom border in unselected tabs

DivineDominion committed:
add height/padding adjustments for title and icon

ChromeStyle.PaddedHeight encapsulated two variants: one has minimum top
padding of the tabs themselves; the other has 70% height of the tab bar.
Icon positions and sizes are adjusted accordingly.

DivineDominion committed:
add Style.tabButtonOffset to enable overlapping buttons

DivineDominion committed:
draw TabButtons with offset so they overlap for Chrome style

DivineDominion committed:
move selected tab to front

onekiloparsec committed:
Merge branch 'master' into chrome-style

# Conflicts:
#	KPCTabsControl.xcodeproj/project.pbxproj
#	KPCTabsControlDemo/ViewController.swift

onekiloparsec committed:
Rounded corner for the Chrome style. But!…

.. the constants not being right, the tabs aren’t really Chrome-like yet.

onekiloparsec commented:
I've fixed a few things that were offset with the `master` branch, and I've added some code to round the corner of the tabs. But the constants aren't good, so the tabs looks too much "vertical" to be Chrome-like. 

I agree that's WIP code! Oh and selecting the tab has no effect on its drawn color?

DivineDominion commented:
For reference purposes:
![chrome](https://cloud.githubusercontent.com/assets/59080/17647672/6d0eb66a-61fb-11e6-98c2-215a5cc22a1b.png)

Selecting a tab should work indeed. What happens on your machine? Here's a screen capture of my commit from yesterday:
![2016-08-14_chrome-tabs](https://cloud.githubusercontent.com/assets/59080/17647673/6d111b6c-61fb-11e6-84b3-f676992a8fea.gif)

Your changes seem to work, too, in respect to highlighting:
![screen shot 2016-08-14 at 08 42 44](https://cloud.githubusercontent.com/assets/59080/17647671/6d0c3052-61fb-11e6-830e-d56a7d90a798.png)

DivineDominion committed:
adjust rounding of ChromeStyle tabs

DivineDominion commented on KPCTabsControl/ChromeStyle.swift:
```diff
@@ -137,47 +137,49 @@ public struct ChromeStyle: Style {
 
         // TODO: Remove these awful hardcoded values...
-        // Remember that Origin start in upper left corner.
+        // Remember that Origin start in lower left corner.
```
I think this was source of an error in the arc at the bottom where you ended up with `y-5=(-5)`. I found 5pt to be a bit too much, at least for smaller sizes, but we can play with this some more and make the value depend on the height of the tabs, for example.

onekiloparsec replied:
You're right, I mixed things up with iOS.

onekiloparsec committed:
h: 21->34

onekiloparsec commented:
Selecting a tab indeed work, I don't know what happened yesterday. As for the Chrome style, your Chrome tabs are very much square, aren't they? Here is a shot of mines.
<img width="380" alt="screen shot 2016-08-14 at 09 36 30" src="https://cloud.githubusercontent.com/assets/226224/17647843/e05f20c6-6202-11e6-995b-20212dda2b8d.png">

onekiloparsec committed:
Moving colors of Chrome Style into ChromeTheme + some refactor

onekiloparsec commented:
I've moved all the colors stuff of the Chrome style into its own Theme (whose implementation follow the clever one you wrote in the first place). So now even ChromeStyle fits pretty well into the API design.

I noticed that the selected tab doesn't remain selected after being dragged and moved to another tab place.

DivineDominion committed:
preserve selection state

DivineDominion commented:
That's strange, indeed yours is less edgy!

The selection gets lost after dragging since forever, it seems. I'm working on it today.

DivineDominion committed:
extend CollectionType so someArray[safe: index] works

DivineDominion committed:
fix retaining selection after dragging

Only selectedItem is the authority to determine which button is
selected now; the other values are derived from it.

Changing the selection updates button states and triggers a re-layout.

DivineDominion commented on KPCTabsControl/TabsControl.swift:
```diff
@@ -461,12 +464,24 @@ public class TabsControl: NSControl, TabEditingDelegate {
+
+    private func updateButtonStatesForSelection() {
+        for button in self.tabButtons() {
+            if let equatableSelectedItem = selectedItem as? String,
```
We may have a problem here: Swift bridges stuff like `String` to `AnyObject` via `NSString` perhaps, but under the hood the equality checks fail. It's not very promising to support every struct from Foundation here. Casting to `Equatable` doesn't work, either.

My suggestions: 

a. Either introduce a `TabButtonItem` protocol which sports an equality check. People who use strings can extend the `String` type on the client side.
b. Work with indexes instead. Selection changes the index to a new value; dragging updates the index when swapping.

(I kind of favor (b)).

onekiloparsec replied:
I thought instance equality would work. But I see that it can be misleading if it is a kind of easy-to-produce volatile thing such as a literal string. Working with Indexes sounds good to me too (b).

DivineDominion replied:
Okay then I'll roll with indexes 👍

DivineDominion committed:
clean up code, adding index to TabButton instead of using tag

DivineDominion committed:
fix window restoration not using the proper selection mechanism

DivineDominion committed:
change selection to be based on button index

DivineDominion commented:
Was easier than I thought. It seems to be working so far!
- [x] Dragging a tab from the middle to the beginning makes the first button reset its frame without animation (because of reloadTabs)

---

Update: moved to #16

DivineDominion committed:
clean up notifying delegate about new item order

onekiloparsec committed:
Using a sensible model class.

onekiloparsec committed:
Removed obsolete method

onekiloparsec committed:
Demonstrating the alternative icon

DivineDominion committed:
fix Chrome title not cutting off right offset

DivineDominion committed:
request field editor style from Style.editorSettings()

DivineDominion committed:
fix half-pixel anti-alialising in NumbersStyle

DivineDominion committed:
retain selection after editing

DivineDominion committed:
fix sub-pixel differences in ChromeStyle

DivineDominion commented:
Found another issue: when invoking the field editor for ChromeStyle tabs with left alignment, there's a 3px horizontal inset and the text moves up 1px it seems. So if you double-click to edit the title seems to move. That's a weird one, though, and playing with offsets here and there didn't help at all.

`NSText`/`NSTextView` come embedded in a `_NSKeyboardClipView` which seems to take care of the clipping and scrolling when you type stuff that's longer than the text field. But fiddling around with that didn't help, either. Ideas?

DivineDominion committed:
fix vertical line drawn on sub-pixels

DivineDominion committed:
fix bottom curve math, increasing radius to 4

DivineDominion committed:
adjust protocol documentation syntax to SwiftDoc

DivineDominion committed:
add selected item to delegate callback

DivineDominion committed:
notify delegate about selection from TabsControl

onekiloparsec commented:
This is all great work. Chrome tabs looks good now. Time to update the list in #9 isn't it?

DivineDominion committed:
ensure the tab field editor is editable

DivineDominion committed:
invoke title change callback even when the string is empty

DivineDominion committed:
remove Theme dependency from Style to allow unthemed user styles

DivineDominion committed:
publicize Offset convenience initializers

DivineDominion committed:
expose NSPoint+Offset math to clients

DivineDominion committed:
expose Theme initializers to client code

Without an empty initializer, it works from the same target; but client
code will not have access to this "implicit" initializer by default.

DivineDominion commented on KPCTabsControl/DefaultTheme.swift:
```diff
@@ -13,7 +13,6 @@ import Cocoa
  */
 public struct DefaultTheme: Theme {
 
-    // TODO: Why is this necessary?
```
Without an empty initializer, it works from the same target; but client code will not have access to this "implicit" initializer by default.

onekiloparsec committed:
Small refactor. Trying to achieve a more consistent naming

DivineDominion committed:
use ChromeTheme font/color settings consistently

DivineDominion committed:
expose Themed protocol so the demo app can access the BG color again

DivineDominion commented:
I like shorter file names 👍 

Found why the field editor's font looked a bit odd: the ChromeStyle didn't use a consistent font setting :) I still don't understand why it is 3px off to the right, though

DivineDominion committed:
make TabsControl non-opaque to allow vibrant views look through

onekiloparsec commented:
Moving the theme away from the Style protocol produce a additional one-line-only protocol, and a not-used typealias `ThemedStyle`. And the point was only to make the theme... optional. It would make sense to make the var `theme` back to `Style` protocol but as an optional. Don't you think?

onekiloparsec committed:
Licence line

onekiloparsec committed:
Using self. to be consistent

onekiloparsec committed:
Some documentation

onekiloparsec committed:
Tweak

onekiloparsec committed:
Added some visibility modifiers, but I am not entirely sure

onekiloparsec committed:
Removing unused stuff

This was probably temporary anyway

onekiloparsec committed:
Using ‘title’ makes more sense with the rest of the API

onekiloparsec committed:
Entirely removing the ‘highlight’ stuff

This can be implemented on the client side anyway, and was conflicting with the highlighted property anyway for a button etc.

onekiloparsec committed:
Oups

onekiloparsec committed:
Removing the Highlighting thing from Theme

DivineDominion commented:
Good stuff 👏

If we put `theme` into `Style` as an optional, client themes would still have to declare the attribute even though their style doesn't use it. That would work, but it's a bit muddy. I was thinking about removing both the protocol and the attribute because we currently only use it in the demo setup to adjust the background colors of the bottom pane. Looking back, this was a quick hack at first and somehow still is. _Using_ a theme from client code is totally possible without having to tell anybody that you do it seems. (I integrate the cutting edge version in my app to play around with the setup.)

Can you come up with a requirement to make `theme` a required property in you way more complex projects?

When we settle on this, it seems this branch is about ready to be merged. 👍

That being said, the drag & drop issue remains but I can't come up with a working solution right now, so I've moved this into another issue. It's not pretty but still a cosmetic problem.

DivineDominion commented on KPCTabsControl/DefaultTheme.swift:
```diff
@@ -26,15 +26,15 @@ public struct DefaultTheme: Theme {
     }
 
-    struct HighlightedTabButtonTheme: KPCTabsControl.TabButtonTheme {
+    private struct HighlightedTabButtonTheme: KPCTabsControl.TabButtonTheme {
```
can we get rid of this variant completely, since there's no highlighting anymore?

onekiloparsec replied:
It's already commited and pushed yesterday evening normally. I'll check on my side but... have a look in yours. :-)
Cédric

Sent from an iPhone 6s running iOS9 at 8 kiloparsecs from the Galactic Center

> Le 18 août 2016 à 10:04, Christian Tietze notifications@github.com a écrit :
> 
> In KPCTabsControl/DefaultTheme.swift:
> 
> > ```
> >      var backgroundColor: NSColor { return DefaultTheme.sharedBackgroundColor }
> >      var borderColor: NSColor { return DefaultTheme.sharedBorderColor }
> >      var titleColor: NSColor { return NSColor.darkGrayColor() }
> >      var titleFont: NSFont { return NSFont.systemFontOfSize(13) }
> >  }
> > ```
> > -    struct HighlightedTabButtonTheme: KPCTabsControl.TabButtonTheme {
> > -    private struct HighlightedTabButtonTheme: KPCTabsControl.TabButtonTheme {
> >   can we get rid of this variant completely, since there's no highlighting anymore?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

DivineDominion replied:
Sure, nevermind, missed that :)

onekiloparsec committed:
Removed Themed protocol and reintegrating theme var into Style

onekiloparsec committed:
Better encapsulation of various TabButton elements

But does the TabEditing thing remain meaningful?

DivineDominion commented:
Love the change to move `finishEditing` to `TabButton`! 👍

`TabEditing` takes hold of references during editing temporarily. I didn't come up with a better way to implement the callbacks without moving more and more logic into the control itself. (Although that's still an option.)

I'm not too happy about the optional `Theme` property because now we're force-unwrapping that all the time. There's no upside of integrating it into `Style` but there's one downside for every force-unwrapping we have to do, and as I mentioned it's now more cumbersome for clients to write Styles if they don't need themes.

onekiloparsec commented:
I agree the control itself contains a lot of logic. The editing as it is now is quite a bit of code for not too much gain, even if I totally agree with the overall approach. I'm currently trying to see what it looks like all back to the control.

My global feeling that we approach a point where we are fairly happy with the features, but not entirely with the API. As you said, we are close to merge the PR, and then we can still work to streamline the APIs, and maybe extract more from the control in a more useful way.

As for the forced unwrapping of the theme, I agree, but isn't it that bad? My feeling isn't definitive either. We'll see.

onekiloparsec commented:
I couldn't wrap my head around that `theme` thing. Could you explain a bit more why such `theme` should be optional (i.e. having a `Style` with no `theme`)? I mean, if we provide one by default, what's the point?

onekiloparsec committed:
Getting rid of TabEditing + fixes of editing process

especially when ending the editing

DivineDominion committed:
less force-unwrapping

DivineDominion commented:
Regarding `theme`: the point is that users will _have_ to implement the optional property. Now even though they use their own means to draw the bezels and text, and thus don't want to use any existing theme's colors, the property lies around pretty useless.

Now this would make sense if any other object had a dependency on `theme`, because then the library _needs_ `theme` to be present. But currently, only the concrete `Style` implementations use the theme. It's an implementation detail of the current set of styles. It's not necessary to use the type (i.e. from another part of the library).

> I mean, if we provide one by default, what's the point?

That's an argument targeting the costs: clients can use the built-in themes and everything's fine. But if they don't want to, they still have to drag the property around.

onekiloparsec committed:
Using tuple

onekiloparsec committed:
Trying to fix issue #16

But I don’t understand how. :-)

onekiloparsec committed:
Better use of delegate

onekiloparsec committed:
Made again theme-less Style and…

And created a specialized Style  as ThemedStyle, since this is what we actually mean, right? The original themed style is a specialisation of the more generic Style.

onekiloparsec commented:
You were perfectly right for the `theme` thing. I needed some time (and sleep) to wrap my head around it. I removed the optional theme from the Style, and made a specialisation of the Style protocol: a `ThemedStyle` protocol, since, I think, this is what we actually try to mean. A `ThemedStyle` is a style using a theme, and has, thus no real meaning to live floating in space alone by itself. Better be the specialisation of `Style`. 

I am gonna merge the request for now, since the `ChromeStyle` itself is now fairly stable. And we can keep going on finishing the 2.0 APi. Thanks for the great conversation and work!

DivineDominion commented:
I was torn between subclassing Style or not, but your argument makes sense. There probably won't be anything else which can use a trait/micin called "Themed" anyway 👍

Also: 🎉 Chrome tabs are done 🎉 :) Now to the other tasks ...
